server/src/main/java/io/crate/expression/AbstractFunctionModule.java

```diff
@@ -34,10 +38,16 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
 import java.util.function.BiFunction;
+import java.util.stream.Collectors;
 
 public abstract class AbstractFunctionModule<T extends FunctionImplementation> extends AbstractModule {
 
+    private static ConcurrentMap<Integer, Signature> SYSTEM_FUNCTION_SIGNATURES_BY_OID = new ConcurrentHashMap<>();
```

Comment:
Having this as global static kinda defeats the purpose of using guice/injection for the functions.

I think this should move into `Functions`. Then the `evaluate` of a `Scalar` somehow needs to get access to it, or to a subset / narrower interface.

I'm also not sure if it is worth to pre-populate a full map given the use case. I suppose `pg_function_is_visible` won't be called too often, so we could consider doing the lookup and oid calculation on the fly and trade increased memory footprint with a higher compute demand when it is actually used.